<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Catherine Benavides
- Ayudante: Nicolás Ojeda, Eduardo Moya

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Tomás Apablaza Arriagada
- Nombre de alumno 2: Maximiliano Quintero Ortega


### **Link de repositorio de GitHub:** `https://github.com/Repositivo/Repositivo.git`

### Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 7 días desde la publicación, 3 días de atraso con 1 punto de descuento c/u.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos Y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquer material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

### Librerias utiles para el Lab 😸

In [117]:
%pip install fsspec

Note: you may need to restart the kernel to use updated packages.


In [118]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
!pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

# 1. Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos.
El objetivo principal será el de realizar un trabajo de limpieza de prepración y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.




In [119]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = 'Dirección donde tiene los archivos en el Drive'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```





In [120]:
df1_grades = pd.read_json("C://Users//maxit//OneDrive//Escritorio//Lab 3//202404051051A8E5909D200AAD03__students_grades.json")
#df2_grades = pd.read_json("C://Users//maxit//OneDrive//Escritorio//Lab 3//202404051051A366AEE5701A1EF2__students_grades_1.json")
#df3_grades = pd.read_json("C://Users//maxit//OneDrive//Escritorio//Lab 3//202404051051565376D77F2A149F__students_grades_2.json")

#df_grades = pd.concat([df1_grades,df2_grades,df3_grades])

df_grades = pd.DataFrame(df1_grades)

print(df_grades)

df1_grades

                 names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1      Madeline Fuller  female        group C            some high school   
2        Daniel Rester    male        group E          associate's degree   
3     Charles Linstrom    male        group A           bachelor's degree   
4       Geraldine Hale  female        group A          associate's degree   
...                ...     ...            ...                         ...   
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
0   

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1      Madeline Fuller  female        group C            some high school   
2        Daniel Rester    male        group E          associate's degree   
3     Charles Linstrom    male        group A           bachelor's degree   
4       Geraldine Hale  female        group A          associate's degree   
...                ...     ...            ...                         ...   
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
0         standard                    none         37            46   
1         standard                    none        124           142   
2         standard                    none        156           196   
3         standard               completed         80            78   
4         standard                    none        131           166   
...            ...                     ...        ...           ...   
1395      standard                    none        193           142   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1398  free/reduced                    none        193           196   
1399      standard               completed         68            80   

     writing score  
0               46  
1              142  
2              193  
3               81  
4              166  
...            ...  
1395           193  
1396            81  
1397            62  
1398           193  
1399            76  

[1400 rows x 9 columns]


<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **875** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **875** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier alumno que posea un valor diferente a esta escala, debería ser eliminado de la base de datos.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?


10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [121]:
def exploratory_data_analysis(dataframe):
    # 1) imprimir dimensiones
    print("Estas son las dimensiones del dataframe:", dataframe.shape)
    # 2) Imprimir las columnas
    print("Estas son las columnas del dataframe:", dataframe.columns.to_list())
    # 3.1) Realizar muestreo de los 5 primeros elementos
    print("Estas son las 5 primeras filas del dataframe:") 
    display(dataframe.head(5))
    # 3.2) Realizar muestreo de los 5 últimos elementos
    print("Estas son las 5 últimos filas del dataframe:") 
    display(dataframe.tail(5))
    # 3.3) Realizar muestreo de los 5 elementos aleatorios
    print("Estas son 5 filas aleatorias del dataframe:")
    display(dataframe.sample(5))
    # 4) Descripción de variables numéricas
    print("Descripción de variables numéricas del dataframe:")
    display(dataframe.describe())
    # 5) Mostrar cantidad de valores nulos
    print("N° de valores nulos del dataframe:", dataframe.isna().sum())
    # 6) Mostrar cantidad de valores únicos
    print("N° de valores únicos del dataframe:", dataframe.nunique())
    # 7) Mostrar cantidad de filas duplicadas
    print("N° de filas duplicadas del dataframe:", dataframe.duplicated().sum())
    # 8) Mostrar filas duplicadas
    print("N° de filas duplicadas del dataframe:")
    display(dataframe.duplicated(subset=["names"]))
    pass


In [122]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df_grades.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

**Justificación:**

```
9) Algunas anomalias, es que el numero de registros es de 1400 registros, no obstante, el numero de nombres unicos son 1153, por lo que se repiten nombres . Por otro lado en la variable test preparation course se observa que el valor más frecuente es None, con una cantidad de 898 registros, por lo que más de la mitad de los alumnos no rindio este test ( considerando cada registro único). Tambéin, pareciese que la escala de las notas no tiene un estandar definido, es decir, aparecen notas de arriba de 100 y notas con numeros menores a 10 y decimales. Además no existen valores nulos, y si existen filas duplicadas.

10) El 50% de los datos en los 3 test, se encuentra entre 100 y 55 aprox., luego el bigote del boxplot se desplaza hasta llegar a 150 y aqui se encuentra un 25% de de los datos ( entre el 100- 150). Luego por otro lado, el 25% de los datos restante se encuentra debajo de 50 y la mediana en todos los casos cercana a 70 . Finalmente existen outliers en los 3 test, mayoritariamente sobre los 150 puntos.
```

**Respuesta Esperada:**

In [123]:
exploratory_data_analysis(df_grades)

Estas son las dimensiones del dataframe: (1400, 9)
Estas son las columnas del dataframe: ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']
Estas son las 5 primeras filas del dataframe:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course math score reading score writing score  
0  standard                    none         37            46            46  
1  standard                    none        124           142           142  
2  standard                    none        156           196           193  
3  standard               completed         80            78            81  
4  standard                    none        131           166           166

Estas son las 5 últimos filas del dataframe:


names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
1395      standard                    none        193           142   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1398  free/reduced                    none        193           196   
1399      standard               completed         68            80   

     writing score  
1395           193  
1396            81  
1397            62  
1398           193  
1399            76

Estas son 5 filas aleatorias del dataframe:


names gender race/ethnicity parental level of education  \
834     Lance Russell   male        group D                 high school   
1298  Clarence Querry   male        group C          associate's degree   
23         Eric Brown   male        group C          associate's degree   
1355    Zachary Glenn   male        group E                some college   
879   Cameron Estrada   male        group D            some high school   

             lunch test preparation course math score reading score  \
834       standard                    none         57            50   
1298      standard                    none         92            79   
23    free/reduced               completed        124           142   
1355      standard               completed         99            87   
879       standard                    none        193                 

     writing score  
834             54  
1298            84  
23             124  
1355            81  
879

Descripción de variables numéricas del dataframe:


names  gender race/ethnicity parental level of education  \
count             1400    1400           1400                        1400   
unique            1153       2              5                           6   
top     Robert Frazier  female        group C                some college   
freq                 3     723            452                         334   

           lunch test preparation course  math score  reading score  \
count       1400                    1400        1400           1400   
unique         2                       2          97             89   
top     standard                    none         193            193   
freq         904                     898         120            117   

        writing score  
count            1400  
unique             93  
top               193  
freq              133

N° de valores nulos del dataframe: names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64
N° de valores únicos del dataframe: names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       97
reading score                    89
writing score                    93
dtype: int64
N° de filas duplicadas del dataframe: 3
N° de filas duplicadas del dataframe:


0       False
1       False
2       False
3       False
4       False
        ...  
1395     True
1396    False
1397    False
1398     True
1399    False
Length: 1400, dtype: bool

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y cree una función que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**

- [ ] Crear función que permita limpiar el dataset.
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hint**

Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.


**Respuesta Esperada:**

In [124]:
def clean_data(dataframe):
    # Transformar a entero las notas
    dataframe["math score"] = pd.to_numeric(dataframe["math score"], errors="coerce")
    dataframe["reading score"] = pd.to_numeric(dataframe["reading score"], errors="coerce")
    dataframe["writing score"] = pd.to_numeric(dataframe["writing score"], errors="coerce")

    # Eliminar notas que no esten en el rango de 0 a 100
    dataframe = dataframe.loc[(dataframe["math score"] >= 0) & (dataframe["math score"] <= 100)]
    dataframe = dataframe.loc[(dataframe["reading score"] >= 0) & (dataframe["reading score"] <= 100)]
    dataframe = dataframe.loc[(dataframe["writing score"] >= 0) & (dataframe["writing score"] <= 100)]
    return dataframe

    # Eliminar nombres duplicados, ya que no pueden dar los mismos exámenes dos veces
    dataframe = dataframe.drop_duplicates(subset=["names"])

    # Eliminar filas duplicadas en general
    dataframe = dataframe.drop_duplicates()

    # Eliminar nulos
    dataframe = dataframe.dropna()

    return dataframe

df_grades = clean_data(df_grades)
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
3         standard               completed        80.0           78.0   
5         standard                    none        73.0           66.0   
6         standard               completed        81.0           75.0   
7         standard               completed        94.0           86.0   
...            ...                     ...         ...            ...   
1392      standard                    none        69.0           75.0   
1394  free/reduced               completed        23.0           44.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1399      standard               completed        68.0           80.0   

      writing score  
0              46.0  
3              81.0  
5              63.0  
6              76.0  
7              87.0  
...             ...  
1392           71.0  
1394           36.0  
1396           81.0  
1397           62.0  
1399           76.0  

[1003 rows x 9 columns]

> Utilice la función creada en el punto 1.1 para verificar sus resultados.

In [125]:
exploratory_data_analysis(df_grades)

Estas son las dimensiones del dataframe: (1003, 9)
Estas son las columnas del dataframe: ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']
Estas son las 5 primeras filas del dataframe:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
3  standard               completed        80.0           78.0           81.0  
5  standard                    none        73.0           66.0           63.0  
6  standard               completed        81.0           75.0           76.0  
7  standard               completed        94.0           86.0           87.0

Estas son las 5 últimos filas del dataframe:


names  gender race/ethnicity parental level of education  \
1392   Richard Young    male        group D                 high school   
1394   Wanda Russell  female        group B                 high school   
1396  Marina Zeigler  female        group C           bachelor's degree   
1397   Laurie Carter  female        group B            some high school   
1399    Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
1392      standard                    none        69.0           75.0   
1394  free/reduced               completed        23.0           44.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1399      standard               completed        68.0           80.0   

      writing score  
1392           71.0  
1394           36.0  
1396           81.0  
1397           62.0  
1399           76.0

Estas son 5 filas aleatorias del dataframe:


names  gender race/ethnicity parental level of education  \
321     Nancy Bassett  female        group C                 high school   
455   Osvaldo Barraza    male        group D            some high school   
921    Agatha Bergman  female        group D             master's degree   
1184   William Musick    male        group B                some college   
820     Nicole Obrien  female        group B           bachelor's degree   

             lunch test preparation course  math score  reading score  \
321       standard                    none        54.0           59.0   
455       standard                    none        86.0           80.0   
921       standard                    none        92.0          100.0   
1184      standard                    none        69.0           54.0   
820   free/reduced                    none        77.0           85.0   

      writing score  
321            62.0  
455            75.0  
921           100.0  
1184           55.0  
820            87.0

Descripción de variables numéricas del dataframe:


math score  reading score  writing score
count  1003.000000    1003.000000    1003.000000
mean     65.996012      69.057827      67.937188
std      15.236983      14.724233      15.326553
min       0.000000      17.000000      10.000000
25%      56.500000      59.000000      57.000000
50%      66.000000      70.000000      69.000000
75%      77.000000      79.000000      79.000000
max     100.000000     100.000000     100.000000

N° de valores nulos del dataframe: names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64
N° de valores únicos del dataframe: names                          1000
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       81
reading score                    72
writing score                    77
dtype: int64
N° de filas duplicadas del dataframe: 3
N° de filas duplicadas del dataframe:


0       False
3       False
5       False
6       False
7       False
        ...  
1392    False
1394    False
1396    False
1397    False
1399    False
Length: 1003, dtype: bool

**Justificación:**

```
Los números únicos de nombres se redujieron considerablemente, al igual que el numero de filas totales. Por otro lado las escalas de de notas de los estudiantes que estan sobre 100 fueron eliminadas y por ende, como se ve en describe() el valor máximo es 100. 
```

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [126]:
def nota_chilena(dataframe_in, columna):
  dataframe_in[columna] = (((dataframe_in[columna])*0.06) +1)
  dataframe_in[columna] = dataframe_in[columna].round(2)
  return dataframe_in

**Respuesta esperada:**

In [127]:
df_grades = nota_chilena(df_grades, ["math score","reading score","writing score"])

display(df_grades)

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score  
0              3.76  
3              5.86  
5              4.78  
6              5.56  
7              6.22  
...             ...  
1392           5.26  
1394           3.16  
1396           5.86  
1397           4.72  
1399           5.56  

[1003 rows x 9 columns]

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [128]:
def alumnos_promedio(dataframe_in):
  dataframe_in["promedio"] = dataframe_in[["math score","reading score","writing score"]].mean(axis=1).round(2)
  return dataframe_in

**Respuesta Esperada:**

In [129]:
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none        3.22           3.76           3.76   
3  standard               completed        5.80           5.68           5.86   
5  standard                    none        5.38           4.96           4.78   
6  standard               completed        5.86           5.50           5.56   
7  standard               completed        6.64           6.16           6.22   

   promedio  
0      3.58  
3      5.78  
5      5.04  
6      5.64  
7      6.34

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta:**

In [130]:
def reprobados_destacados(dataframe_in):
    dataframe1 = dataframe_in.loc[dataframe_in['promedio'] < 4]
    dataframe2 = dataframe_in.copy()
    dataframe2.loc[dataframe2['promedio'] < 4, 'promedio'] = 'R'
    dataframe3 = dataframe_in[(dataframe_in['math score'] > 5.8) & (dataframe_in['writing score'] > 5.8) & (dataframe_in['reading score'] > 5.8)]
    dataframe3 = dataframe3.sort_values(by='promedio', ascending=False)
    return dataframe1, dataframe2, dataframe3


In [131]:
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)
dataframe1

C:\Users\maxit\AppData\Local\Temp\ipykernel_19852\3589467954.py:4: FutureWarning:

Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'R' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.



names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
13        Mikki Amaya  female        group A                 high school   
26      Samuel Dougal    male        group C             master's degree   
55    Rodger Talavera    male        group B                 high school   
63     Morris Johnson    male        group A          associate's degree   
...               ...     ...            ...                         ...   
1371   Charles Harian    male        group C            some high school   
1373    Wendy Gunnell  female        group D            some high school   
1389      Diana Costa  female        group C                 high school   
1391      Amy Sieving  female        group C                 high school   
1394    Wanda Russell  female        group B                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
13    free/reduced               completed        3.04           3.88   
26    free/reduced               completed        3.76           3.52   
55    free/reduced                    none        3.16           2.74   
63    free/reduced               completed        3.40           4.30   
...            ...                     ...         ...            ...   
1371  free/reduced               completed        3.70           4.12   
1373  free/reduced                    none        2.62           3.04   
1389  free/reduced                    none        3.46           3.76   
1391      standard                    none        2.74           2.74   
1394  free/reduced               completed        2.38           3.64   

      writing score  promedio  
0              3.76      3.58  
13             3.46      3.46  
26             3.76      3.68  
55             2.62      2.84  
63             4.18      3.96  
...             ...       ...  
1371           3.94      3.92  
1373           2.92      2.86  
1389           3.58      3.60  
1391           2.80      2.76  
1394           3.16      3.06  

[106 rows x 10 columns]

In [132]:
dataframe2

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score promedio  
0              3.76        R  
3              5.86     5.78  
5              4.78     5.04  
6              5.56     5.64  
7              6.22     6.34  
...             ...      ...  
1392           5.26      5.3  
1394           3.16        R  
1396           5.86     5.42  
1397           4.72     4.54  
1399           5.56     5.48  

[1003 rows x 10 columns]

In [133]:
dataframe3

names  gender race/ethnicity parental level of education  \
1352      Sandra Pompey  female        group E          associate's degree   
384   Elizabeth Beasley  female        group E           bachelor's degree   
677       Richard Pryor    male        group E           bachelor's degree   
475           Emma Gray  female        group E           bachelor's degree   
1276     Angela Cressey  female        group D                some college   
...                 ...     ...            ...                         ...   
949      Rolando Widmer    male        group C                 high school   
296   Lucille Stanphill  female        group C                 high school   
300        Steven Kelly    male        group D                some college   
1307        Peter Perez    male        group D             master's degree   
96          Harry Berry    male        group B          associate's degree   

         lunch test preparation course  math score  reading score  \
1352  standard                    none        7.00           7.00   
384   standard                    none        7.00           7.00   
677   standard               completed        7.00           7.00   
475   standard               completed        6.94           7.00   
1276  standard                    none        6.88           7.00   
...        ...                     ...         ...            ...   
949   standard               completed        5.92           6.04   
296   standard                    none        5.86           6.04   
300   standard                    none        5.86           5.92   
1307  standard                    none        5.86           5.86   
96    standard               completed        5.86           5.92   

      writing score  promedio  
1352           7.00      7.00  
384            7.00      7.00  
677            7.00      7.00  
475            7.00      6.98  
1276           6.94      6.94  
...             ...       ...  
949            5.92      5.96  
296            5.92      5.94  
300            6.04      5.94  
1307           6.04      5.92  
96             5.92      5.90  

[110 rows x 10 columns]

**Respuesta:**

```
Escriba aqui su respuesta
```

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [134]:
# Código para quintiles
# Define las etiquetas personalizadas
etiquetas = ['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']


GPA_cuantiles = pd.qcut(df_grades["promedio"], 5, labels=etiquetas)


df_grades['GPA cuantiles'] = GPA_cuantiles

print(df_grades)
"""
NOTA: La variable del dataframe que contenga los quintiles debe llamarse
'GPA cuantiles' para que el gráfico funcione.
"""

                 names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0 

"\nNOTA: La variable del dataframe que contenga los quintiles debe llamarse\n'GPA cuantiles' para que el gráfico funcione.\n"

In [135]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_grades.sort_values(by="promedio"),
    x="promedio",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig


```
No existe una relación muy clara, los cortes de los quintiles casi siempre terminan y comienzan donde mismo, en las categorias de educación de los padres. Solo se podría decir que en el quintil "Bajisimo" se concentra más en la nota 4 a mayor nivel educacional de los padree, sin embargo solo puede ser concecuencia de la dispersión de los datos producto de que existen menos datos de padres con master y los demás estan mucho mpas dispersos. 
```

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

In [137]:
df2_grades = pd.read_json("C://Users//maxit//OneDrive//Escritorio//Lab 3//202404051051A366AEE5701A1EF2__students_grades_1.json")
df3_grades = pd.read_json("C://Users//maxit//OneDrive//Escritorio//Lab 3//202404051051565376D77F2A149F__students_grades_2.json")

df_grades_nueva = pd.concat([df2_grades,df3_grades])
df_grades_nueva

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
470     Richard Young    male        group D                 high school   
471     Wanda Russell  female        group B                 high school   
472    Marina Zeigler  female        group C           bachelor's degree   
473     Laurie Carter  female        group B            some high school   
474      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
470      standard                    none        5.14           5.50   
471  free/reduced               completed        2.38           3.64   
472  free/reduced               completed        4.96           5.44   
473      standard               completed        4.24           4.66   
474      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
470           5.26  
471           3.16  
472           5.86  
473           4.72  
474           5.56  

[875 rows x 9 columns]

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos] **


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


In [140]:
df_grades_nueva["GPA"] = df_grades_nueva[["math score","reading score","writing score"]].mean(axis=1).round(2)


education_mapping = {
    'some high school': 'school',
    'some college': 'school',
    'high school': 'school',
    "bachelor's degree": 'college',
    "associate's degree": 'college',
    "master's degree": 'postgraduate'
}
df_grades_nueva['simple parental level of education'] = df_grades_nueva['parental level of education'].map(education_mapping)

# Paso 3: Agregar según 2 niveles: race/ethnicity y simple parental level of education
summary_table = df_grades_nueva.groupby(['race/ethnicity', 'simple parental level of education']).agg(
    GPA=('GPA', 'mean'),
    count=('GPA', 'count')
).reset_index()

total_counts = df_grades_nueva.groupby('race/ethnicity')['GPA'].count().reset_index()
total_counts.columns = ['race/ethnicity', 'count']
summary_table = pd.concat([summary_table, total_counts])

total_students = df_grades_nueva['GPA'].count()
summary_table['porcentaje'] = (summary_table['count'] / total_students * 100).map("{:.2f}%".format)

display(summary_table)

race/ethnicity simple parental level of education       GPA  count  \
0         group A                            college  4.876667     24   
1         group A                       postgraduate  5.090000      2   
2         group A                             school  4.617647     51   
3         group B                            college  5.175556     54   
4         group B                       postgraduate  5.384000      5   
5         group B                             school  4.780374    107   
6         group C                            college  5.249020    102   
7         group C                       postgraduate  5.057333     15   
8         group C                             school  4.902581    155   
9         group D                            college  5.201714     70   
10        group D                       postgraduate  5.498000     20   
11        group D                             school  5.085906    149   
12        group E                            college  5.482308     52   
13        group E                       postgraduate  5.820000      6   
14        group E                             school  5.288889     63   
0         group A                                NaN       NaN     77   
1         group B                                NaN       NaN    166   
2         group C                                NaN       NaN    272   
3         group D                                NaN       NaN    239   
4         group E                                NaN       NaN    121   

   porcentaje  
0       2.74%  
1       0.23%  
2       5.83%  
3       6.17%  
4       0.57%  
5      12.23%  
6      11.66%  
7       1.71%  
8      17.71%  
9       8.00%  
10      2.29%  
11     17.03%  
12      5.94%  
13      0.69%  
14      7.20%  
0       8.80%  
1      18.97%  
2      31.09%  
3      27.31%  
4      13.83%

In [142]:
fig = px.box(
    df_grades_nueva,
    x="simple parental level of education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
)
fig.show()

**Resultado Esperado de la tabla**

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.74 |            5    |            4.89 |  4.88 |      24 | 2.74 %       |
|  1 |           | postgraduate                         |         4.69 |            5.23 |            5.35 |  5.09 |       2 | 0.23 %       |
|  2 |           | school                               |         4.57 |            4.73 |            4.56 |  4.62 |      51 | 5.83 %       |
|  3 | group B          | college                              |         5.07 |            5.26 |            5.19 |  5.18 |      54 | 6.17 %       |
|  4 |           | postgraduate                         |         4.91 |            5.69 |            5.55 |  5.38 |       5 | 0.57 %       |
|  5 |           | school                               |         4.69 |            4.89 |            4.76 |  4.78 |     107 | 12.23 %      |
|  6 | group C          | college                              |         5.02 |            5.37 |            5.35 |  5.25 |     102 | 11.66 %      |
|  7 |           | postgraduate                         |         4.92 |            5.14 |            5.1  |  5.06 |      15 | 1.71 %       |
|  8 |           | school                               |         4.76 |            5.02 |            4.92 |  4.9  |     155 | 17.71 %      |
|  9 | group D          | college                              |         5.11 |            5.25 |            5.25 |  5.2  |      70 | 8.0 %        |
| 10 |           | postgraduate                         |         5.22 |            5.54 |            5.73 |  5.5  |      20 | 2.29 %       |
| 11 |           | school                               |         5.02 |            5.13 |            5.11 |  5.09 |     149 | 17.03 %      |
| 12 | group E          | college                              |         5.54 |            5.45 |            5.45 |  5.48 |      52 | 5.94 %       |
| 13 |          | postgraduate                         |         5.54 |            6.03 |            5.89 |  5.82 |       6 | 0.69 %       |
| 14 |          | school                               |         5.4  |            5.31 |            5.16 |  5.29 |      63 | 7.2 %        |

### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene nuevamente los alumnos de los registros corruptos que ud. y su equipo filtraron en el análisis anterior.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

In [144]:
other_grades = pd.read_csv('C://Users//maxit//OneDrive//Escritorio//Lab 3//2024040510517433FE38528DF944__other_grades.csv')

In [146]:

outer_join = pd.merge(df_grades, other_grades, on='names', how='outer')

outer_join

names  gender race/ethnicity parental level of education  \
0         Aaron Oakden    male        group D                 high school   
1         Aaron Oakden    male        group D                 high school   
2         Aaron Shelly    male        group D          associate's degree   
3         Aaron Shelly    male        group D          associate's degree   
4          Adam Romero    male        group C                 high school   
...                ...     ...            ...                         ...   
1873       Willie Neff  female        group E          associate's degree   
1874       Willie Neff  female        group E          associate's degree   
1875  Willie Nicholson  female        group D                some college   
1876  Willie Nicholson  female        group D                some college   
1877     Zachary Glenn    male        group E                some college   

             lunch test preparation course  math score  reading score  \
0     free/reduced                    none        4.96           5.44   
1     free/reduced                    none        4.96           5.44   
2         standard                    none        4.66           3.88   
3         standard                    none        4.66           3.88   
4         standard               completed        4.18           4.12   
...            ...                     ...         ...            ...   
1873  free/reduced                    none        5.38           5.56   
1874  free/reduced                    none        5.38           5.56   
1875  free/reduced               completed        4.78           5.80   
1876  free/reduced               completed        4.78           5.80   
1877      standard               completed        6.94           6.22   

      writing score  promedio GPA cuantiles  science score  history score  
0              5.14      5.18         Medio             46            4.6  
1              5.14       NaN           NaN             46            4.6  
2              3.76      4.10      Bajísimo             48            3.9  
3              3.76       NaN           NaN             48            3.9  
4              3.94      4.08      Bajísimo             42            3.2  
...             ...       ...           ...            ...            ...  
1873           5.68      5.54          Alto             58            4.6  
1874           5.68       NaN           NaN             58            4.6  
1875           5.80      5.46          Alto             63            4.8  
1876           5.80       NaN           NaN             63            4.8  
1877           5.86      6.34      Muy alto             99            5.7  

[1878 rows x 13 columns]

In [147]:
left_join = pd.merge(df_grades, other_grades, on='names', how='left')
left_join


names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1873     Richard Young    male        group D                 high school   
1874     Wanda Russell  female        group B                 high school   
1875    Marina Zeigler  female        group C           bachelor's degree   
1876     Laurie Carter  female        group B            some high school   
1877      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
1         standard               completed        5.80           5.68   
2         standard                    none        5.38           4.96   
3         standard               completed        5.86           5.50   
4         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1873      standard                    none        5.14           5.50   
1874  free/reduced               completed        2.38           3.64   
1875  free/reduced               completed        4.96           5.44   
1876      standard               completed        4.24           4.66   
1877      standard               completed        5.08           5.80   

      writing score  promedio GPA cuantiles  science score  history score  
0              3.76      3.58      Bajísimo             25            3.8  
1              5.86      5.78          Alto             56            5.2  
2              4.78      5.04         Medio             58            3.8  
3              5.56      5.64          Alto             72            5.0  
4              6.22      6.34      Muy alto             94            4.6  
...             ...       ...           ...            ...            ...  
1873           5.26       NaN           NaN             62            4.2  
1874           3.16       NaN           NaN             16            3.6  
1875           5.86       NaN           NaN             59            4.1  
1876           4.72       NaN           NaN             43            4.3  
1877           5.56       NaN           NaN             61            5.3  

[1878 rows x 13 columns]

In [148]:
right_join = pd.merge(df_grades, other_grades, on='names', how='right')

right_join

names  gender race/ethnicity parental level of education  \
0            Pam Otoole  female        group B           bachelor's degree   
1            Pam Otoole  female        group B           bachelor's degree   
2           Diane Olson  female        group C                some college   
3           Diane Olson  female        group C                some college   
4        Sarah Flannery  female        group B             master's degree   
...                 ...     ...            ...                         ...   
1873  Margaret Spannaus  female        group C                 high school   
1874     Vickie Sellers  female        group D                some college   
1875     Vickie Sellers  female        group D                some college   
1876      Kari Mitchell  female        group D                some college   
1877      Kari Mitchell  female        group D                some college   

             lunch test preparation course  math score  reading score  \
0         standard                    none        5.32           5.32   
1         standard                    none        5.32           5.32   
2         standard               completed        5.14           6.40   
3         standard               completed        5.14           6.40   
4         standard                    none        6.40           6.70   
...            ...                     ...         ...            ...   
1873  free/reduced               completed        4.54           5.26   
1874      standard               completed        5.08           5.68   
1875      standard               completed        5.08           5.68   
1876  free/reduced                    none        5.62           6.16   
1877  free/reduced                    none        5.62           6.16   

      writing score  promedio GPA cuantiles  science score  history score  
0              5.44      5.36          Alto             57            4.0  
1              5.44       NaN           NaN             57            4.0  
2              6.28      5.94      Muy alto             48            4.8  
3              6.28       NaN           NaN             48            4.8  
4              6.58      6.56      Muy alto             72            6.1  
...             ...       ...           ...            ...            ...  
1873           4.90       NaN           NaN             47            4.0  
1874           5.62      5.46          Alto             47            5.7  
1875           5.62       NaN           NaN             47            5.7  
1876           6.16      5.98      Muy alto             53            6.2  
1877           6.16       NaN           NaN             53            6.2  

[1878 rows x 13 columns]

In [149]:
inner_join = pd.merge(df_grades, other_grades, on='names', how='inner')

inner_join

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1873     Richard Young    male        group D                 high school   
1874     Wanda Russell  female        group B                 high school   
1875    Marina Zeigler  female        group C           bachelor's degree   
1876     Laurie Carter  female        group B            some high school   
1877      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
1         standard               completed        5.80           5.68   
2         standard                    none        5.38           4.96   
3         standard               completed        5.86           5.50   
4         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1873      standard                    none        5.14           5.50   
1874  free/reduced               completed        2.38           3.64   
1875  free/reduced               completed        4.96           5.44   
1876      standard               completed        4.24           4.66   
1877      standard               completed        5.08           5.80   

      writing score  promedio GPA cuantiles  science score  history score  
0              3.76      3.58      Bajísimo             25            3.8  
1              5.86      5.78          Alto             56            5.2  
2              4.78      5.04         Medio             58            3.8  
3              5.56      5.64          Alto             72            5.0  
4              6.22      6.34      Muy alto             94            4.6  
...             ...       ...           ...            ...            ...  
1873           5.26       NaN           NaN             62            4.2  
1874           3.16       NaN           NaN             16            3.6  
1875           5.86       NaN           NaN             59            4.1  
1876           4.72       NaN           NaN             43            4.3  
1877           5.56       NaN           NaN             61            5.3  

[1878 rows x 13 columns]

El join definido será el inner join pues entrega solo los registros que tengan los nombres coincidentes y por tanto no considerra aquellos registros que no estan en una base ni en la otra 

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>